text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+`rng.gen_range(a, b)` can become `rng.gen(a..b)`. However, this makes
+the non-ranged generation slightly more complicated/"magic
+incantation"-y: `rng.gen(..)` (this relies on having explicit syntax
+for `FullRange`, per [RFC 702]).
+
+[RFC 702]: https://github.com/rust-lang/rfcs/pull/702
+
+The convention for using `FullRange` for unrestricted generation may
+seem peculiar for types that don't have any concept of ranges,
+e.g. creating a randomly seeded RNG, but the consistency and
+"vagueness" of `..` likely means this is OK.
+
+(NB. general discussion, non-normative) This approach allows us to
+naturally leverage any "range combinators" used elsewhere,
+e.g. inclusive endpoints, and so remove the need for the `Closed01`
+and `Open01` types. It also allows us to implement `Random` for
+e.g. `Vec`, say, `rng.gen::<Vec<u8>>(10..20)` would create a vector of
+some length between 10 and 20 and fill it with `u8`s.
+
+#### `#[derive]`
+
+The deriving mode will be removed from the compiler, or at least,
+disabled on the stable channel; it can be rarely used since most data
+types have some restriction on their values and/or non-trivial
+relationships between them, which a derived `Random` implementation
+can do nothing to preserve.
+
+#### Alternatives
+
+Call it `Rand` as it is today.
+
+Not introduce the `Data` parameter. This would be simpler for this
+trait, but require duplication elsewhere.
+
+Have either `Self` or `Data` be associated types. This would require
+more proxy types since e.g. the following is illegal:
+
+```rust
+impl Random for u8 {
+    type Data = FullRange;
+    // ...
+}
+impl Random for u8 {
+    type Data = Range<u8>;
+    // ...
+}
+```
+
+### The `Rng` trait
+
+[the `Rng` trait]: #the-rng-trait
+
+This trait represent random number generators, it is mostly provided
+for expressing intent, and for optimisation purposes over a plain
+`Reader`.
+
+It is expected and recommended that types implementing `Rng` generates
+a sequence of bits where each bit:
+
+- has equal probability of being either 0 or 1,
+- is independent of all other bits.
+
+Other functions will be assuming these properties to give sensible
+answers (e.g. `shuffle`, `choose` and `sample`) but there's no way any
+function can rely on them and cause unsafety only when violated, so
+this is not overly problematic.
+
+Currently this trait is not object safe, but like `Iterator` and
+`Reader` (etc.) it may be desirable to abstract over many different
+RNG types, so making it object safe is desirable.
+
+The methods provided are all optional, implemented in terms of
+`Reader::read`, e.g. as a sketch (no [error handling])
+
+```rust
+fn read_u32(&mut self) -> u32 {
+    let mut buf = [0u8; 4];
+    self.read(&mut buf);
+
+    ((buf[0] as u32) << 24) |
+    ((buf[1] as u32) << 16) |
+    ((buf[2] as u32) << 8) |
+    ((buf[3] as u32) << 0)
+}
+```
+
+They are provided because most RNG types naturally generate one of
```

Comment:
@nagisa this RFC is actually suggesting that RNGs should implement `Reader` (`Rng` inheriting from it requires it) since the sequence of output values can be regarded as a bit/byte stream. The current `Rng` trait provides `fill_bytes` which is literally `read` (without error handling).

This paragraph is justifying the existence of the `next_` methods, rather than just only using `Reader` for everything.

@aturon could you expand a little about which parts of IO reform?